File: src/converter/parser.ts
/**
 * Minimal JSONata parser covering the mapping subset the converter can
 * translate: object/array constructors, paths, predicates, arithmetic,
 * comparison and boolean operators, `&` concatenation, ternaries, `$fn(...)`
 * calls, and `function($x){...}` lambdas (as arguments). Anything outside the
 * subset raises a ConvertError with the offending construct.
 */

export class ConvertError extends Error {}

export type JNode =
  | { kind: 'num'; value: number }
  | { kind: 'str'; value: string }
  | { kind: 'bool'; value: boolean }
  | { kind: 'null' }
  | { kind: 'name'; value: string }
  | { kind: 'context' } // `$`
  | { kind: 'root' } // `$$`
  | { kind: 'parent'; depth: number } // `%` (depth 1), `%.%` (depth 2), …
  | { kind: 'dot'; left: JNode; right: JNode }
  | { kind: 'predicate'; target: JNode; expr: JNode }
  | { kind: 'call'; name: string; args: JNode[] }
  | { kind: 'lambda'; params: string[]; body: JNode }
  | { kind: 'bind'; target: JNode; variable: string; op: '#' | '@' }
  | { kind: 'binary'; op: string; left: JNode; right: JNode }
  | {
      kind: 'sort';
      target: JNode;
      terms: { key: JNode; descending: boolean }[];
    }
  | {
      kind: 'block';
      bindings: { name: string; value: JNode }[];
      result: JNode;
    }
  | { kind: 'ternary'; cond: JNode; then: JNode; else: JNode }
  | { kind: 'object'; pairs: { key: string; value: JNode }[] }
  | { kind: 'array'; items: JNode[] }
  | { kind: 'neg'; operand: JNode };

interface JToken {
  type: 'num' | 'str' | 'name' | 'var' | 'punct' | 'eof';
  value: string;
}

const PUNCT2 = new Set(['!=', '<=', '>=', ':=', '~>', '**']);
const PUNCT1 = '{}[](),:.?&=<>+-*/%;!@#^|';

function lex(source: string): JToken[] {
  const tokens: JToken[] = [];
  let i = 0;
  const n = source.length;
  while (i < n) {
    const c = source[i]!;
    if (c === ' ' || c === '\t' || c === '\r' || c === '\n') {
      i++;
      continue;
    }
    if (c === '/' && source[i + 1] === '*') {
      const end = source.indexOf('*/', i + 2);
      if (end === -1) throw new ConvertError('Unterminated comment.');
      i = end + 2;
      continue;
    }
    if (/[0-9]/.test(c)) {
      let j = i;
      while (j < n && /[0-9.eE+-]/.test(source[j]!)) {
        // stop at operators that only look like number parts
        if (
          (source[j] === '+' || source[j] === '-') &&
          !/[eE]/.test(source[j - 1]!)
        )
          break;
        j++;
      }
      const raw = source.slice(i, j);
      const value = Number.parseFloat(raw);
      tokens.push({ type: 'num', value: String(value) });
      i += String(raw).length;
      continue;
    }
    if (c === '"' || c === "'") {
      let j = i + 1;
      let value = '';
      while (j < n && source[j] !== c) {
        if (source[j] === '\\') {
          value += source[j + 1] ?? '';
          j += 2;
        } else {
          value += source[j]!;
          j++;
        }
      }
      if (j >= n) throw new ConvertError('Unterminated string literal.');
      tokens.push({ type: 'str', value });
      i = j + 1;
      continue;
    }
    if (c === '`') {
      const end = source.indexOf('`', i + 1);
      if (end === -1) throw new ConvertError('Unterminated quoted name.');
      tokens.push({ type: 'name', value: source.slice(i + 1, end) });
      i = end + 1;
      continue;
    }
    if (c === '$') {
      let j = i + 1;
      if (source[j] === '$') j++;
      while (j < n && /[A-Za-z0-9_]/.test(source[j]!)) j++;
      tokens.push({ type: 'var', value: source.slice(i, j) });
      i = j;
      continue;
    }
    if (/[A-Za-z_]/.test(c)) {
      let j = i;
      while (j < n && /[A-Za-z0-9_]/.test(source[j]!)) j++;
      tokens.push({ type: 'name', value: source.slice(i, j) });
      i = j;
      continue;
    }
    const two = source.slice(i, i + 2);
    if (PUNCT2.has(two)) {
      tokens.push({ type: 'punct', value: two });
      i += 2;
      continue;
    }
    if (PUNCT1.includes(c)) {
      tokens.push({ type: 'punct', value: c });
      i++;
      continue;
    }
    throw new ConvertError(`Unexpected character '${c}'.`);
  }
  tokens.push({ type: 'eof', value: '' });
  return tokens;
}

class JParser {
  private pos = 0;
  constructor(private tokens: JToken[]) {}

  private peek(): JToken {
    return this.tokens[this.pos]!;
  }
  private next(): JToken {
    const t = this.peek();
    if (t.type !== 'eof') this.pos++;
    return t;
  }
  private at(value: string): boolean {
    const t = this.peek();
    return (t.type === 'punct' || t.type === 'name') && t.value === value;
  }
  private eat(value: string): boolean {
    if (this.at(value)) {
      this.pos++;
      return true;
    }
    return false;
  }
  private expect(value: string): void {
    if (!this.eat(value)) {
      throw new ConvertError(
        `Expected '${value}' but found '${this.peek().value || 'end of input'}'.`,
      );
    }
  }

  parse(): JNode {
    const expr = this.parseExpr();
    if (this.peek().type !== 'eof') {
      throw new ConvertError(
        `Unsupported JSONata construct near '${this.peek().value}'.`,
      );
    }
    return expr;
  }

  parseExpr(): JNode {
    return this.parseTernary();
  }

  private parseTernary(): JNode {
    const cond = this.parseOr();
    if (!this.eat('?')) return cond;
    const then = this.parseExpr();
    this.expect(':');
    const els = this.parseExpr();
    // oxlint-disable-next-line unicorn/no-thenable -- AST data field, never awaited
    return { kind: 'ternary', cond, then, else: els };
  }

  private parseBinary(ops: string[], nextLevel: () => JNode): JNode {
    let left = nextLevel.call(this);
    for (;;) {
      const t = this.peek();
      const isWordOp = t.type === 'name' && ops.includes(t.value);
      const isPunctOp = t.type === 'punct' && ops.includes(t.value);
      if (!isWordOp && !isPunctOp) return left;
      this.next();
      const right = nextLevel.call(this);
      left = { kind: 'binary', op: t.value, left, right };
    }
  }

  private parseOr(): JNode {
    return this.parseBinary(['or'], this.parseAnd);
  }
  private parseAnd(): JNode {
    return this.parseBinary(['and'], this.parseComparison);
  }
  private parseComparison(): JNode {
    return this.parseBinary(
      ['=', '!=', '<', '<=', '>', '>=', 'in'],
      this.parseConcat,
    );
  }
  private parseConcat(): JNode {
    return this.parseBinary(['&'], this.parseAdditive);
  }
  private parseAdditive(): JNode {
    return this.parseBinary(['+', '-'], this.parseMultiplicative);
  }
  private parseMultiplicative(): JNode {
    return this.parseBinary(['*', '/', '%'], this.parseUnary);
  }

  private parseUnary(): JNode {
    if (this.eat('-')) {
      return { kind: 'neg', operand: this.parseUnary() };
    }
    return this.parsePostfix();
  }

  private parsePostfix(): JNode {
    let expr = this.parsePrimary();
    for (;;) {
      if (this.at('.')) {
        this.next();
        if (this.at('{')) {
          // `arr.{ ... }` — per-element object constructor (projection).
          expr = { kind: 'dot', left: expr, right: this.parseObject() };
          continue;
        }
        if (this.at('(')) {
          // `arr.( ... )` — per-element mapping of a block/expression; the
          // parenthesized form allows `$var := ...` bindings before the result.
          expr = { kind: 'dot', left: expr, right: this.parsePrimary() };
          continue;
        }
        const t = this.peek();
        if (t.type === 'punct' && t.value === '%') {
          // `%.%`, `%.%.%`, … — a multi-level parent. Accumulate the depth.
          if (expr.kind !== 'parent') {
            throw new ConvertError('`.%` is only valid after a `%` parent.');
          }
          this.next();
          expr = { kind: 'parent', depth: expr.depth + 1 };
          continue;
        }
        if (t.type !== 'name') {
          throw new ConvertError(
            `Unsupported path step after '.': '${t.value}'.`,
          );
        }
        this.next();
        expr = {
          kind: 'dot',
          left: expr,
          right: { kind: 'name', value: t.value },
        };
      } else if (this.at('[')) {
        this.next();
        const inner = this.parseExpr();
        this.expect(']');
        expr = { kind: 'predicate', target: expr, expr: inner };
      } else if (this.at('^')) {
        // JSONata order-by: `arr^(>a, <b)` — `>` descending, `<` (or bare) ascending.
        this.next();
        this.expect('(');
        const terms: { key: JNode; descending: boolean }[] = [];
        while (!this.at(')')) {
          let descending = false;
          if (this.eat('>')) descending = true;
          else this.eat('<');
          terms.push({ key: this.parseExpr(), descending });
          if (!this.eat(',')) break;
        }
        this.expect(')');
        if (terms.length === 0) {
          throw new ConvertError('A sort `^(...)` needs at least one key.');
        }
        expr = { kind: 'sort', target: expr, terms };
      } else if (this.at('#') || this.at('@')) {
        // `path#$i` / `path@$v` — positional / context variable binding.
        const op = this.next().value as '#' | '@';
        const t = this.peek();
        if (t.type !== 'var') {
          throw new ConvertError(`Expected a $variable after '${op}'.`);
        }
        this.next();
        expr = { kind: 'bind', target: expr, variable: t.value, op };
      } else {
        return expr;
      }
    }
  }

  private parseObject(): JNode {
    this.expect('{');
    const pairs: { key: string; value: JNode }[] = [];
    while (!this.at('}')) {
      const keyTok = this.peek();
      if (keyTok.type !== 'str' && keyTok.type !== 'name') {
        throw new ConvertError(
          'Only literal object keys are supported by the converter.',
        );
      }
      this.next();
      this.expect(':');
      const value = this.parseExpr();
      pairs.push({ key: keyTok.value, value });
      if (!this.eat(',')) break;
    }
    this.expect('}');
    return { kind: 'object', pairs };
  }

  private parsePrimary(): JNode {
    const t = this.peek();
    if (t.type === 'num') {
      this.next();
      return { kind: 'num', value: Number(t.value) };
    }
    if (t.type === 'str') {
      this.next();
      return { kind: 'str', value: t.value };
    }
    if (t.type === 'var') {
      this.next();
      if (t.value === '$') return { kind: 'context' };
      if (t.value === '$$') return { kind: 'root' };
      // `$name(...)` is a function call; a bare `$name` is a variable (lambda param).
      if (this.at('(')) {
        this.next();
        const args: JNode[] = [];
        while (!this.at(')')) {
          args.push(this.parseArgument());
          if (!this.eat(',')) break;
        }
        this.expect(')');
        return { kind: 'call', name: t.value.slice(1), args };
      }
      return { kind: 'name', value: t.value };
    }
    if (t.type === 'name') {
      if (t.value === 'true' || t.value === 'false') {
        this.next();
        return { kind: 'bool', value: t.value === 'true' };
      }
      if (t.value === 'null') {
        this.next();
        return { kind: 'null' };
      }
      if (t.value === 'function') {
        return this.parseLambda();
      }
      this.next();
      return { kind: 'name', value: t.value };
    }
    if (this.at('(')) {
      this.next();
      // `( $x := e; ...; result )` — a block of variable bindings ending in a
      // result expression. A plain `( expr )` (no bindings) is just grouping.
      const bindings: { name: string; value: JNode }[] = [];
      let result: JNode;
      for (;;) {
        const e = this.parseExpr();
        if (this.eat(':=')) {
          if (e.kind !== 'name' || !e.value.startsWith('$')) {
            throw new ConvertError(
              'The left side of `:=` must be a $variable.',
            );
          }
          bindings.push({ name: e.value, value: this.parseExpr() });
          if (this.eat(';')) continue;
          throw new ConvertError(
            'A `( ... )` block must end with a result expression after its `$var := ...` bindings.',
          );
        }
        if (this.at(';')) {
          throw new ConvertError(
            'Only `$var := ...` statements are supported before the final expression in a `( ... )` block.',
          );
        }
        result = e;
        break;
      }
      this.expect(')');
      if (bindings.length === 0) return result;
      return { kind: 'block', bindings, result };
    }
    // `%` in operand position is JSONata's parent operator (distinct from the
    // infix `%` modulo, which parseBinary consumes between two operands).
    if (this.at('%')) {
      this.next();
      return { kind: 'parent', depth: 1 };
    }
    if (this.at('{')) return this.parseObject();
    if (this.at('[')) {
      this.next();
      const items: JNode[] = [];
      while (!this.at(']')) {
        items.push(this.parseExpr());
        if (!this.eat(',')) break;
      }
      this.expect(']');
      return { kind: 'array', items };
    }
    throw new ConvertError(
      `Unsupported JSONata construct near '${t.value || 'end of input'}'.`,
    );
  }

  private parseArgument(): JNode {
    if (this.peek().type === 'name' && this.peek().value === 'function') {
      return this.parseLambda();
    }
    return this.parseExpr();
  }

  private parseLambda(): JNode {
    this.expect('function');
    this.expect('(');
    const params: string[] = [];
    while (!this.at(')')) {
      const t = this.peek();
      if (t.type !== 'var') {
        throw new ConvertError('Lambda parameters must be $variables.');
      }
      this.next();
      params.push(t.value);
      if (!this.eat(',')) break;
    }
    this.expect(')');
    this.expect('{');
    const body = this.parseExpr();
    this.expect('}');
    return { kind: 'lambda', params, body };
  }
}

export function parseJsonata(source: string): JNode {
  return new JParser(lex(source)).parse();
}

File: apps/converters/src/jsonata/parser.rs
//! Minimal JSONata parser (port of `src/converter/jsonata/parser.ts`):
//! object/array constructors, paths, predicates, arithmetic, comparison and
//! boolean operators, `&` concatenation, ternaries, `$fn(...)` calls, and
//! `function($x){...}` lambdas as arguments.

use crate::util::{js_num, js_parse_float};

#[derive(Clone, Debug)]
pub enum JNode {
    Num(f64),
    Str(String),
    Bool(bool),
    Null,
    Name(String),
    /// `$`
    Context,
    /// `$$`
    Root,
    /// `%` (depth 1), `%.%` (depth 2), …
    Parent {
        depth: usize,
    },
    Dot {
        left: Box<JNode>,
        right: Box<JNode>,
    },
    Predicate {
        target: Box<JNode>,
        expr: Box<JNode>,
    },
    Call {
        name: String,
        args: Vec<JNode>,
    },
    Lambda {
        params: Vec<String>,
        body: Box<JNode>,
    },
    Bind {
        target: Box<JNode>,
        variable: String,
        op: char, // '#' | '@'
    },
    Binary {
        op: String,
        left: Box<JNode>,
        right: Box<JNode>,
    },
    Sort {
        target: Box<JNode>,
        terms: Vec<(JNode, bool)>, // (key, descending)
    },
    Block {
        bindings: Vec<(String, JNode)>,
        result: Box<JNode>,
    },
    Ternary {
        cond: Box<JNode>,
        then: Box<JNode>,
        els: Box<JNode>,
    },
    Object {
        pairs: Vec<(String, JNode)>,
    },
    Array {
        items: Vec<JNode>,
    },
    Neg {
        operand: Box<JNode>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
enum JTokenType {
    Num,
    Str,
    Name,
    Var,
    Punct,
    Eof,
}

#[derive(Clone, Debug)]
struct JToken {
    type_: JTokenType,
    value: String,
}

const PUNCT2: &[&str] = &["!=", "<=", ">=", ":=", "~>", "**"];
const PUNCT1: &str = "{}[](),:.?&=<>+-*/%;!@#^|";

type JResult<T> = Result<T, String>;

fn lex(source: &str) -> JResult<Vec<JToken>> {
    let chars: Vec<char> = source.chars().collect();
    let n = chars.len();
    let mut tokens: Vec<JToken> = Vec::new();
    let mut i = 0usize;
    while i < n {
        let c = chars[i];
        if c == ' ' || c == '\t' || c == '\r' || c == '\n' {
            i += 1;
            continue;
        }
        if c == '/' && chars.get(i + 1) == Some(&'*') {
            let mut j = i + 2;
            let mut found = None;
            while j + 1 < n {
                if chars[j] == '*' && chars[j + 1] == '/' {
                    found = Some(j);
                    break;
                }
                j += 1;
            }
            let Some(end) = found else {
                return Err("Unterminated comment.".to_string());
            };
            i = end + 2;
            continue;
        }
        if c.is_ascii_digit() {
            let mut j = i;
            while j < n
                && (chars[j].is_ascii_digit()
                    || chars[j] == '.'
                    || chars[j] == 'e'
                    || chars[j] == 'E'
                    || chars[j] == '+'
                    || chars[j] == '-')
            {
                // stop at operators that only look like number parts
                if (chars[j] == '+' || chars[j] == '-')
                    && !(chars[j - 1] == 'e' || chars[j - 1] == 'E')
                {
                    break;
                }
                j += 1;
            }
            let raw: String = chars[i..j].iter().collect();
            let value = js_parse_float(&raw);
            tokens.push(JToken {
                type_: JTokenType::Num,
                value: js_num(value),
            });
            i += raw.chars().count();
            continue;
        }
        if c == '"' || c == '\'' {
            let mut j = i + 1;
            let mut value = String::new();
            while j < n && chars[j] != c {
                if chars[j] == '\\' {
                    // Faithful port: the TS lexer appends the escaped char verbatim.
                    if let Some(&next) = chars.get(j + 1) {
                        value.push(next);
                    }
                    j += 2;
                } else {
                    value.push(chars[j]);
                    j += 1;
                }
            }
            if j >= n {
                return Err("Unterminated string literal.".to_string());
            }
            tokens.push(JToken {
                type_: JTokenType::Str,
                value,
            });
            i = j + 1;
            continue;
        }
        if c == '`' {
            let mut j = i + 1;
            let mut found = None;
            while j < n {
                if chars[j] == '`' {
                    found = Some(j);
                    break;
                }
                j += 1;
            }
            let Some(end) = found else {
                return Err("Unterminated quoted name.".to_string());
            };
            tokens.push(JToken {
                type_: JTokenType::Name,
                value: chars[i + 1..end].iter().collect(),
            });
            i = end + 1;
            continue;
        }
        if c == '$' {
            let mut j = i + 1;
            if chars.get(j) == Some(&'$') {
                j += 1;
            }
            while j < n && (chars[j].is_ascii_alphanumeric() || chars[j] == '_') {
                j += 1;
            }
            tokens.push(JToken {
                type_: JTokenType::Var,
                value: chars[i..j].iter().collect(),
            });
            i = j;
            continue;
        }
        if c.is_ascii_alphabetic() || c == '_' {
            let mut j = i;
            while j < n && (chars[j].is_ascii_alphanumeric() || chars[j] == '_') {
                j += 1;
            }
            tokens.push(JToken {
                type_: JTokenType::Name,
                value: chars[i..j].iter().collect(),
            });
            i = j;
            continue;
        }
        if i + 1 < n {
            let two: String = chars[i..i + 2].iter().collect();
            if PUNCT2.contains(&two.as_str()) {
                tokens.push(JToken {
                    type_: JTokenType::Punct,
                    value: two,
                });
                i += 2;
                continue;
            }
        }
        if PUNCT1.contains(c) {
            tokens.push(JToken {
                type_: JTokenType::Punct,
                value: c.to_string(),
            });
            i += 1;
            continue;
        }
        return Err(format!("Unexpected character '{}'.", c));
    }
    tokens.push(JToken {
        type_: JTokenType::Eof,
        value: String::new(),
    });
    Ok(tokens)
}

struct JParser {
    tokens: Vec<JToken>,
    pos: usize,
}

impl JParser {
    fn peek(&self) -> &JToken {
        &self.tokens[self.pos]
    }

    fn next(&mut self) -> JToken {
        let t = self.peek().clone();
        if t.type_ != JTokenType::Eof {
            self.pos += 1;
        }
        t
    }

    fn at(&self, value: &str) -> bool {
        let t = self.peek();
        (t.type_ == JTokenType::Punct || t.type_ == JTokenType::Name) && t.value == value
    }

    fn eat(&mut self, value: &str) -> bool {
        if self.at(value) {
            self.pos += 1;
            true
        } else {
            false
        }
    }

    fn expect(&mut self, value: &str) -> JResult<()> {
        if !self.eat(value) {
            let found = if self.peek().value.is_empty() {
                "end of input".to_string()
            } else {
                self.peek().value.clone()
            };
            return Err(format!("Expected '{}' but found '{}'.", value, found));
        }
        Ok(())
    }

    fn parse(&mut self) -> JResult<JNode> {
        let expr = self.parse_expr()?;
        if self.peek().type_ != JTokenType::Eof {
            return Err(format!(
                "Unsupported JSONata construct near '{}'.",
                self.peek().value
            ));
        }
        Ok(expr)
    }

    fn parse_expr(&mut self) -> JResult<JNode> {
        self.parse_ternary()
    }

    fn parse_ternary(&mut self) -> JResult<JNode> {
        let cond = self.parse_binary_level(0)?;
        if !self.eat("?") {
            return Ok(cond);
        }
        let then = self.parse_expr()?;
        self.expect(":")?;
        let els = self.parse_expr()?;
        Ok(JNode::Ternary {
            cond: Box::new(cond),
            then: Box::new(then),
            els: Box::new(els),
        })
    }

    /// or → and → comparison → `&` concat → additive → multiplicative.
    fn parse_binary_level(&mut self, level: usize) -> JResult<JNode> {
        const LEVELS: &[&[&str]] = &[
            &["or"],
            &["and"],
            &["=", "!=", "<", "<=", ">", ">=", "in"],
            &["&"],
            &["+", "-"],
            &["*", "/", "%"],
        ];
        if level >= LEVELS.len() {
            return self.parse_unary();
        }
        let ops = LEVELS[level];
        let mut left = self.parse_binary_level(level + 1)?;
        loop {
            let t = self.peek();
            let is_word_op = t.type_ == JTokenType::Name && ops.contains(&t.value.as_str());
            let is_punct_op = t.type_ == JTokenType::Punct && ops.contains(&t.value.as_str());
            if !is_word_op && !is_punct_op {
                return Ok(left);
            }
            let op = self.next().value;
            let right = self.parse_binary_level(level + 1)?;
            left = JNode::Binary {
                op,
                left: Box::new(left),
                right: Box::new(right),
            };
        }
    }

    fn parse_unary(&mut self) -> JResult<JNode> {
        if self.eat("-") {
            return Ok(JNode::Neg {
                operand: Box::new(self.parse_unary()?),
            });
        }
        self.parse_postfix()
    }

    fn parse_postfix(&mut self) -> JResult<JNode> {
        let mut expr = self.parse_primary()?;
        loop {
            if self.at(".") {
                self.next();
                if self.at("{") {
                    // `arr.{ ... }` — per-element object constructor (projection).
                    let right = self.parse_object()?;
                    expr = JNode::Dot {
                        left: Box::new(expr),
                        right: Box::new(right),
                    };
                    continue;
                }
                if self.at("(") {
                    // `arr.( ... )` — per-element mapping of a block/expression.
                    let right = self.parse_primary()?;
                    expr = JNode::Dot {
                        left: Box::new(expr),
                        right: Box::new(right),
                    };
                    continue;
                }
                let t = self.peek().clone();
                if t.type_ == JTokenType::Punct && t.value == "%" {
                    // `%.%`, `%.%.%`, … — a multi-level parent.
                    let JNode::Parent { depth } = expr else {
                        return Err("`.%` is only valid after a `%` parent.".to_string());
                    };
                    self.next();
                    expr = JNode::Parent { depth: depth + 1 };
                    continue;
                }
                if t.type_ != JTokenType::Name {
                    return Err(format!("Unsupported path step after '.': '{}'.", t.value));
                }
                self.next();
                expr = JNode::Dot {
                    left: Box::new(expr),
                    right: Box::new(JNode::Name(t.value)),
                };
            } else if self.at("[") {
                self.next();
                let inner = self.parse_expr()?;
                self.expect("]")?;
                expr = JNode::Predicate {
                    target: Box::new(expr),
                    expr: Box::new(inner),
                };
            } else if self.at("^") {
                // JSONata order-by: `arr^(>a, <b)`.
                self.next();
                self.expect("(")?;
                let mut terms: Vec<(JNode, bool)> = Vec::new();
                while !self.at(")") {
                    let mut descending = false;
                    if self.eat(">") {
                        descending = true;
                    } else {
                        self.eat("<");
                    }
                    terms.push((self.parse_expr()?, descending));
                    if !self.eat(",") {
                        break;
                    }
                }
                self.expect(")")?;
                if terms.is_empty() {
                    return Err("A sort `^(...)` needs at least one key.".to_string());
                }
                expr = JNode::Sort {
                    target: Box::new(expr),
                    terms,
                };
            } else if self.at("#") || self.at("@") {
                // `path#$i` / `path@$v` — positional / context variable binding.
                let op = self.next().value.chars().next().unwrap();
                let t = self.peek().clone();
                if t.type_ != JTokenType::Var {
                    return Err(format!("Expected a $variable after '{}'.", op));
                }
                self.next();
                expr = JNode::Bind {
                    target: Box::new(expr),
                    variable: t.value,
                    op,
                };
            } else {
                return Ok(expr);
            }
        }
    }

    fn parse_object(&mut self) -> JResult<JNode> {
        self.expect("{")?;
        let mut pairs: Vec<(String, JNode)> = Vec::new();
        while !self.at("}") {
            let key_tok = self.peek().clone();
            if key_tok.type_ != JTokenType::Str && key_tok.type_ != JTokenType::Name {
                return Err("Only literal object keys are supported by the converter.".to_string());
            }
            self.next();
            self.expect(":")?;
            let value = self.parse_expr()?;
            pairs.push((key_tok.value, value));
            if !self.eat(",") {
                break;
            }
        }
        self.expect("}")?;
        Ok(JNode::Object { pairs })
    }

    fn parse_primary(&mut self) -> JResult<JNode> {
        let t = self.peek().clone();
        if t.type_ == JTokenType::Num {
            self.next();
            return Ok(JNode::Num(t.value.parse::<f64>().unwrap_or(f64::NAN)));
        }
        if t.type_ == JTokenType::Str {
            self.next();
            return Ok(JNode::Str(t.value));
        }
        if t.type_ == JTokenType::Var {
            self.next();
            if t.value == "$" {
                return Ok(JNode::Context);
            }
            if t.value == "$$" {
                return Ok(JNode::Root);
            }
            // `$name(...)` is a function call; a bare `$name` is a variable.
            if self.at("(") {
                self.next();
                let mut args: Vec<JNode> = Vec::new();
                while !self.at(")") {
                    args.push(self.parse_argument()?);
                    if !self.eat(",") {
                        break;
                    }
                }
                self.expect(")")?;
                return Ok(JNode::Call {
                    name: t.value[1..].to_string(),
                    args,
                });
            }
            return Ok(JNode::Name(t.value));
        }
        if t.type_ == JTokenType::Name {
            if t.value == "true" || t.value == "false" {
                self.next();
                return Ok(JNode::Bool(t.value == "true"));
            }
            if t.value == "null" {
                self.next();
                return Ok(JNode::Null);
            }
            if t.value == "function" {
                return self.parse_lambda();
            }
            self.next();
            return Ok(JNode::Name(t.value));
        }
        if self.at("(") {
            self.next();
            // `( $x := e; ...; result )` — bindings block; plain `( expr )` is grouping.
            let mut bindings: Vec<(String, JNode)> = Vec::new();
            let result: JNode;
            loop {
                let e = self.parse_expr()?;
                if self.eat(":=") {
                    let name = match &e {
                        JNode::Name(value) if value.starts_with('$') => value.clone(),
                        _ => return Err("The left side of `:=` must be a $variable.".to_string()),
                    };
                    bindings.push((name, self.parse_expr()?));
                    if self.eat(";") {
                        continue;
                    }
                    return Err(
                        "A `( ... )` block must end with a result expression after its `$var := ...` bindings."
                            .to_string(),
                    );
                }
                if self.at(";") {
                    return Err(
                        "Only `$var := ...` statements are supported before the final expression in a `( ... )` block."
                            .to_string(),
                    );
                }
                result = e;
                break;
            }
            self.expect(")")?;
            if bindings.is_empty() {
                return Ok(result);
            }
            return Ok(JNode::Block {
                bindings,
                result: Box::new(result),
            });
        }
        // `%` in operand position is JSONata's parent operator.
        if self.at("%") {
            self.next();
            return Ok(JNode::Parent { depth: 1 });
        }
        if self.at("{") {
            return self.parse_object();
        }
        if self.at("[") {
            self.next();
            let mut items: Vec<JNode> = Vec::new();
            while !self.at("]") {
                items.push(self.parse_expr()?);
                if !self.eat(",") {
                    break;
                }
            }
            self.expect("]")?;
            return Ok(JNode::Array { items });
        }
        let found = if t.value.is_empty() {
            "end of input".to_string()
        } else {
            t.value
        };
        Err(format!("Unsupported JSONata construct near '{}'.", found))
    }

    fn parse_argument(&mut self) -> JResult<JNode> {
        if self.peek().type_ == JTokenType::Name && self.peek().value == "function" {
            return self.parse_lambda();
        }
        self.parse_expr()
    }

    fn parse_lambda(&mut self) -> JResult<JNode> {
        self.expect("function")?;
        self.expect("(")?;
        let mut params: Vec<String> = Vec::new();
        while !self.at(")") {
            let t = self.peek().clone();
            if t.type_ != JTokenType::Var {
                return Err("Lambda parameters must be $variables.".to_string());
            }
            self.next();
            params.push(t.value);
            if !self.eat(",") {
                break;
            }
        }
        self.expect(")")?;
        self.expect("{")?;
        let body = self.parse_expr()?;
        self.expect("}")?;
        Ok(JNode::Lambda {
            params,
            body: Box::new(body),
        })
    }
}

pub fn parse_jsonata(source: &str) -> JResult<JNode> {
    let tokens = lex(source)?;
    JParser { tokens, pos: 0 }.parse()
}
